### PR TITLE
JENKINS-72120 fix discovery and building of tags in 4.0.0

### PR DESCRIPTION
**Pull request by atl-dyon**
Fix for the discovery of tags after upgrading to 4.0.0
Tags will display in the 'tag' tab 

```[tasklist]
### Submitter checklist
- [ ] Make sure you are opening from a **topic/feature/bugfix branch** (right side) and not your main branch!
- [ ] Ensure that the pull request title represents the desired changelog entry
- [ ] Please describe what you did
- [ ] Link to relevant issues in GitHub or Jira
- [ ] Link to relevant pull requests, esp. upstream and downstream changes
- [ ] Ensure you have provided tests - that demonstrates feature works or fixes the issue
```

### CODE REVIEW

File: src/main/java/com/atlassian/bitbucket/jenkins/internal/scm/trait/BitbucketLegacyTraitConverter.java
Comment (mhenschke-atl) [inline src/main/java/com/atlassian/bitbucket/jenkins/internal/scm/trait/BitbucketLegacyTraitConverter.java:27]: So I suppose customers that upgraded to 4.0 before this merges will lose their tag discovery trait and need to create the new one, but those that upgrade after don't need to make any changes? We should communicate this in the release notes and on the bug ticket- this feels like a gotcha.

Comment (atl-dyon) [reply src/main/java/com/atlassian/bitbucket/jenkins/internal/scm/trait/BitbucketLegacyTraitConverter.java:27]: From my understanding they may still need to make changes to their discovery trait as it's a new one. But this is something we can test.

**Comment by S0obi** [comment]
@atl-dyon hello there! Any plan to move forward with this change ? JENKINS-72120 is currently preventing us to update the plugin. Thanks in advance 👍

**Comment by atl-dyon** [comment]
Hi there, still working on this.

File: src/main/java/com/atlassian/bitbucket/jenkins/internal/client/BitbucketRepositoryClient.java
Comment (atl-dyon) [inline src/main/java/com/atlassian/bitbucket/jenkins/internal/client/BitbucketRepositoryClient.java:121]: `4.1.0`

File: src/main/java/com/atlassian/bitbucket/jenkins/internal/client/BitbucketTagClient.java
Comment (dkjellin) [inline src/main/java/com/atlassian/bitbucket/jenkins/internal/client/BitbucketTagClient.java:7]: some javadoc please, both here and on the method.

File: src/main/java/com/atlassian/bitbucket/jenkins/internal/model/BitbucketDefaultBranch.java
Comment (dkjellin) [inline src/main/java/com/atlassian/bitbucket/jenkins/internal/model/BitbucketDefaultBranch.java:22]: why this change?
Are we not getting the correct `type` coming through?

File: src/main/java/com/atlassian/bitbucket/jenkins/internal/scm/BitbucketSCMSource.java
Comment (dkjellin) [inline src/main/java/com/atlassian/bitbucket/jenkins/internal/scm/BitbucketSCMSource.java:641]: why this change? (not a problem just trying to understand)

**Review by dkjellin — COMMENTED** [review]
Overall looks good to me, just some minor things and I think we're good to go.
I also think you should change the version in the `pom.ml` to `4.1.0-SNAPSHOT` and add an entry to the ReadMe.md with a reference to the ticket and saying we implemented it.

File: src/main/java/com/atlassian/bitbucket/jenkins/internal/model/BitbucketDefaultBranch.java
Comment (atl-dyon) [reply src/main/java/com/atlassian/bitbucket/jenkins/internal/model/BitbucketDefaultBranch.java:22]: It's been a bit since i've looked at this. I'll double check

File: src/main/java/com/atlassian/bitbucket/jenkins/internal/scm/BitbucketSCMSource.java
Comment (atl-dyon) [reply src/main/java/com/atlassian/bitbucket/jenkins/internal/scm/BitbucketSCMSource.java:641]: old accidental holdover from when i added tags as a default trait for some testing. Will change back

**Comment by DStefanow** [comment]
Hello, I also wonder, are you going to implement 'auto build of tags' with the discover? Because currently I am testing on one dev instance with latest Jenkins and only the discover works auto start of build does not.

File: docs/tag_discovery.md
Comment (mhenschke-atl) [inline docs/tag_discovery.md:14]: sadf

File: src/main/java/com/atlassian/bitbucket/jenkins/internal/scm/BitbucketTagSCMHead.java
Comment (atl-dyon) [reply src/main/java/com/atlassian/bitbucket/jenkins/internal/scm/BitbucketTagSCMHead.java:47]: Yeah this was an intentional change to save just copying BranchRevision line for line, maybe its worth changing it to just an SCMRevision as both branch and tag require identical revision classes.

File: src/main/java/com/atlassian/bitbucket/jenkins/internal/model/BitbucketDefaultBranch.java
Comment (atl-dyon) [reply src/main/java/com/atlassian/bitbucket/jenkins/internal/model/BitbucketDefaultBranch.java]: This may have been a temporary change when debugging some tag issues. Reverted to previous implementation.

File: src/main/java/com/atlassian/bitbucket/jenkins/internal/scm/BitbucketTagSCMHead.java
Comment (atl-dyon) [reply src/main/java/com/atlassian/bitbucket/jenkins/internal/scm/BitbucketTagSCMHead.java:47]: scratch above, was thinking about something else. Have updated it!

File: docs/tag_discovery.md
Comment (mhenschke-atl) [inline docs/tag_discovery.md:7]: @jaguilar-atl might explain this better but I wonder if we need this comment? Other behaviours that do this exist, so I don't know if we specifically need to provide one.

File: pom.xml
Comment (mhenschke-atl) [inline pom.xml:12]: Update the acceptance test pom as well, please

File: docs/tag_discovery.md
Comment (atl-dyon) [reply docs/tag_discovery.md:7]: Was thinking it might be better to exclusively call it out, as we may immediately get bug reports for it without conveying it first.

File: src/main/java/com/atlassian/bitbucket/jenkins/internal/scm/trait/BitbucketLegacyTraitConverter.java
Comment (mhenschke-atl) [reply src/main/java/com/atlassian/bitbucket/jenkins/internal/scm/trait/BitbucketLegacyTraitConverter.java:27]: I think that's worth doing- maybe as part of the blitz?

File: readme.md
Comment (mhenschke-atl) [inline readme.md:228]: Love it 👍

File: src/main/java/com/atlassian/bitbucket/jenkins/internal/client/BitbucketTagClientImpl.java
Comment (mhenschke-atl) [inline src/main/java/com/atlassian/bitbucket/jenkins/internal/client/BitbucketTagClientImpl.java:66]: I don't think you need this- `NextPageFetcherImpl` has private access to it's containing class so you should be able to reference `MAX_PAGES`

File: src/main/java/com/atlassian/bitbucket/jenkins/internal/model/BitbucketTag.java
Comment (mhenschke-atl) [inline src/main/java/com/atlassian/bitbucket/jenkins/internal/model/BitbucketTag.java:15]: /nit formatting here is a bit wonky

File: src/main/java/com/atlassian/bitbucket/jenkins/internal/model/BitbucketTag.java
Comment (mhenschke-atl) [inline src/main/java/com/atlassian/bitbucket/jenkins/internal/model/BitbucketTag.java:30]: I think we should consider `latestCommit` here as well?

File: src/main/java/com/atlassian/bitbucket/jenkins/internal/client/BitbucketTagClientImpl.java
Comment (mhenschke-atl) [inline src/main/java/com/atlassian/bitbucket/jenkins/internal/client/BitbucketTagClientImpl.java:88]:
```suggestion
                taskListener.getLogger().println("Tag retrieval limit reached.");
```

**Comment by mhenschke-atl** [comment]
I'm going to conditionally approve this PR- I think the approach looks good, like the limits on tag retrieval, but I'd like to hold off on merging until the team has done some manual upgrade testing from 4.x and 3.x (and some more general testing- pushing those page limits, tag deletion etc.)

There aren't acceptance tests for this work which I'd like to do in the future, but it should not block this PR. Once the testing is done we should merge this and angle for a 4.1.0 release ASAP.

Great work on this.

**Review by mhenschke-atl — APPROVED** [review]

File: src/main/java/com/atlassian/bitbucket/jenkins/internal/client/BitbucketTagClientImpl.java
Comment (atl-dyon) [reply src/main/java/com/atlassian/bitbucket/jenkins/internal/client/BitbucketTagClientImpl.java:66]: We do it this way in branchClient and PRClient and are warned that we need a synthetic accessor.

Comment (mhenschke-atl) [reply src/main/java/com/atlassian/bitbucket/jenkins/internal/client/BitbucketTagClientImpl.java:66]: Is there any reason a synthetic accessor is a concern here?

Comment (jaguilar-atl) [reply src/main/java/com/atlassian/bitbucket/jenkins/internal/client/BitbucketTagClientImpl.java:66]: I think it was mostly done to make testing unit testing easier by being able to use different mock values.

File: src/main/java/com/atlassian/bitbucket/jenkins/internal/trigger/BitbucketWebhookConsumer.java
Comment (mhenschke-atl) [inline src/main/java/com/atlassian/bitbucket/jenkins/internal/trigger/BitbucketWebhookConsumer.java:303]: This looks good to me 👍

File: src/main/java/com/atlassian/bitbucket/jenkins/internal/client/BitbucketTagClientImpl.java
Comment (mhenschke-atl) [inline src/main/java/com/atlassian/bitbucket/jenkins/internal/client/BitbucketTagClientImpl.java:48]: Big comment

File: src/main/java/com/atlassian/bitbucket/jenkins/internal/client/BitbucketTagClientImpl.java
Comment (mhenschke-atl) [inline src/main/java/com/atlassian/bitbucket/jenkins/internal/client/BitbucketTagClientImpl.java:103]: cmoment

File: src/main/java/com/atlassian/bitbucket/jenkins/internal/client/BitbucketTagClientImpl.java
Comment (mhenschke-atl) [inline src/main/java/com/atlassian/bitbucket/jenkins/internal/client/BitbucketTagClientImpl.java:37]: comment

File: src/main/java/com/atlassian/bitbucket/jenkins/internal/status/LocalSCMListener.java
Comment (mhenschke-atl) [inline src/main/java/com/atlassian/bitbucket/jenkins/internal/status/LocalSCMListener.java:140]: asdf

File: src/main/java/com/atlassian/bitbucket/jenkins/internal/scm/BitbucketSCMSource.java
Comment (mhenschke-atl) [inline src/main/java/com/atlassian/bitbucket/jenkins/internal/scm/BitbucketSCMSource.java:308]: Thanks for fixing the tag SCM head. The two above are still incorrect- out of the scope for this PR but can we add a comment like
```code
                    // JENKINS-73267 fetch commit from source rather than cached value
```

File: src/main/java/com/atlassian/bitbucket/jenkins/internal/scm/BitbucketSCMSource.java
Comment (mhenschke-atl) [inline src/main/java/com/atlassian/bitbucket/jenkins/internal/scm/BitbucketSCMSource.java:312]: Does the `BitbucketTagSCMHead` persist the latest commit in this code, and what happens if someone pushes up a change without a webhook, then tries to build this job?
I don't think we can trust this value- this is the "last _seen_ commit", not the latest commit. I have no opposition to persisting it, but wanted to see if we should have this conditional or always do the fetch.

Comment (atl-dyon) [reply src/main/java/com/atlassian/bitbucket/jenkins/internal/scm/BitbucketSCMSource.java:312]: If we dont have a commit for the tag, fetch it, then push and build again without webhooks nothing will change. Tags are stuck on the commit they were created on. So we never have to fetch again and can trust it once we have a value.

Comment (mhenschke-atl) [reply src/main/java/com/atlassian/bitbucket/jenkins/internal/scm/BitbucketSCMSource.java:313]: > Tags are stuck on the commit they were created on.

You can force push a different commit to a tag with the same name to Bitbucket. It's generally not great, but there are legitimate cases for doing so (burning a release as part of a `mvn release:prepare` comes immediately to mind).

I tried this out locally- make a branch with commits `a` and `b`, tag `a` as newtag and push & build it in Jenkins, then `git tag -f newtag` with `b` checked out, force push it up, and build again- Jenkins will continue to build `a`.

Tags _can_ change, but I don't think they change very often, and there's a good reason why Jenkins doesn't automatically build them so my gut feel is the slight inefficiency of checking every time, even though 99% of the time it will be the same is probably worth not having this issue- if you do want to rebuild a tag with a different commit, you either have to push up a different tag or edit it in UI which isn't great.

Comment (atl-dyon) [reply src/main/java/com/atlassian/bitbucket/jenkins/internal/scm/BitbucketSCMSource.java]: All fixed.

File: src/main/java/com/atlassian/bitbucket/jenkins/internal/model/BitbucketTag.java
Comment (dkjellin) [inline src/main/java/com/atlassian/bitbucket/jenkins/internal/model/BitbucketTag.java:40]:
```suggestion
   @CheckForNull
    public String getLatestCommit() {
```

File: src/main/java/com/atlassian/bitbucket/jenkins/internal/scm/BitbucketSCMSource.java
Comment (dkjellin) [inline src/main/java/com/atlassian/bitbucket/jenkins/internal/scm/BitbucketSCMSource.java:484]: This is a blind get. which has me a bit nervous. A quick and dirty fix would be to change the return type to be `Optional<BitbucketTag>` and then change this method to

```java
BitbucketSCMSource.DescriptorImpl descriptor = (BitbucketSCMSource.DescriptorImpl) getDescriptor();
        return descriptor.getConfiguration(getServerId()).map(serverConfiguration -> {BitbucketScmHelper scmHelper = descriptor.getBitbucketScmHelper(serverConfiguration.getBaseUrl(), getCredentials().orElse(null));
            return scmHelper.getTagClient(getProjectKey(), getRepositorySlug(), listener)
                    .getRemoteTag(head.getName());});
```

File: src/main/java/com/atlassian/bitbucket/jenkins/internal/scm/BitbucketSCMSource.java
Comment (dkjellin) [inline src/main/java/com/atlassian/bitbucket/jenkins/internal/scm/BitbucketSCMSource.java:317]: If you go with my suggestion below, you can then change this to
```java
Optional<BitbucketTag> fetchedTag = fetchBitbucketTag((BitbucketTagSCMHead) head, listener);
            return new BitbucketSCMRevision((BitbucketTagSCMHead) head, fetchedTag.map(BitbucketTag::getLatestCommit).orElse(null));
```

of course deciding if `null` or empty String is the correct default.

**Review by dkjellin — APPROVED** [review]
I think this is good. The blind get on the optional should probably be fixed but I would argue we fix that in a separate smaller PR. The javadoc I commented on would also benefit from updating, but again that can be done after this is merged. 

I did not go over the tests in my review, I know others have done so, so I will opt out of reviewing them this time :D

File: src/main/java/com/atlassian/bitbucket/jenkins/internal/client/BitbucketTagClient.java
Comment (atl-dyon) [reply src/main/java/com/atlassian/bitbucket/jenkins/internal/client/BitbucketTagClient.java:34]: Good catch. I chose not to include the throws as we dont in the BitbucketBranchClient & we'd be exposing some pretty obscure exceptions that typically shouldn't be encountered. Like the `BitbucketClientException` & `NoContentException`

**Pull request merged**
